runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -100,6 +103,40 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         val stream = conn.makeRequest(engineRequest, respHandler)
         stream.activate()
 
+        if (request.isChunked) {
+            val bytesToRead = 65536L
+
+            when (request.body) {
+                is HttpBody.SourceContent -> {
+                    val source = (request.body as HttpBody.SourceContent).readFrom()
+                    val nextBuffer = SdkBuffer()
+                    var buffer = SdkBuffer()
+
+                    while (true) {
+                        source.read(buffer, bytesToRead)
+
+                        // read another set of bytes into a second buffer -- this is how we can "peek" to decide if this is the final chunk or not
+                        val isFinalChunk = source.read(nextBuffer, bytesToRead) == -1L
+
+                        stream.writeChunk(buffer.readToByteArray(), isFinalChunk)
+
+                        if (isFinalChunk) break else buffer = nextBuffer
+                    }
+                }
+                is HttpBody.ChannelContent -> {
+                    val chan = (request.body as HttpBody.ChannelContent).readFrom()
+                    val buffer = SdkBuffer()
+
+                    while (!chan.isClosedForRead) {
+                        val bytesRead = chan.read(buffer, bytesToRead)
+                        check(bytesRead != -1L) { "unexpected exhaustion of channel" }
```

Comment:
We can't send a 0-byte chunk as the final chunk, it has to have some data. This is because CRT does a check on the size of the chunk and skips processing if it's 0.

I see the issue, I will refactor it